### ISSUE TITLE
Soft character insert does not trigger brace highlighting

### ISSUE THREAD

**Issue by ervumlens** [Komodo/KomodoEdit#395]
Just a minor inconsistency. Not hurting me any.

**Problem**
The auto-inserted close character (e.g., `]`) and its matching open character (e.g., `[`) are not highlighted when soft characters are enabled.

**Steps to Reproduce**
1. Enable soft characters in Preferences.
2. Open a file and type `[`.

**Expected**
The closing `]` is automatically inserted. Both braces are highlighted.
![image](https://cloud.githubusercontent.com/assets/13005338/8509918/6206258e-228a-11e5-9dd6-694eb6b7f0fb.png)

**Actual**
The closing `]` is automatically inserted, but neither brace is highlighted. Note that moving the cursor out of the braces and back in will trigger highlighting.

![image](https://cloud.githubusercontent.com/assets/13005338/8509913/0134489e-228a-11e5-901b-47075b563b34.png)

**Comment by cgchoffman** [comment]
Some small additional details.  I almost closed this as it appeared to be by design:
The initial indicator state of the softchar appears to be incorrect.  As shown in the second image above, the initial state is underline indicator state is underlined and blue.

The actually indicator should be just underlined.  You can see this if you simply add a character between the brackets eg. "[b]".  "]" will be underlined and white like it's partner.  I suspect this is related to #396 and indicators being left behind on typed characters (seen by holding down a key, typed letter will be highlighted until key is lifted, then softchar is highlighted again)

**Comment by ervumlens** [comment]
> The initial indicator state of the softchar appears to be incorrect.

Good catch. 

Apparently the softchar and its underline are inserted after the regular styling pass. That pass includes looking for matching braces (e.g., [the call to `doMatchBrace`](https://github.com/Komodo/KomodoEdit/blob/8865f7249de2715c648b6aebc37b0f00b6466214/src/chrome/komodo/content/bindings/views-buffer.p.xml#L1109), which [is also a delayed call](https://github.com/Komodo/KomodoEdit/blob/8865f7249de2715c648b6aebc37b0f00b6466214/src/chrome/komodo/content/bindings/views-buffer.p.xml#L1169) as described in #396 ). 

Here's a quick proof. If a softchar `]` is auto-inserted too late for matching, what happens when there is a floating `]` sitting there to start with?

1. Type `]`
2. Move the cursor to the left of the `]`
3. Type `[`

![image](https://cloud.githubusercontent.com/assets/13005338/8535708/fd29a656-240d-11e5-95f8-592c8f139e37.png)


Result: The softchar pair are styled correctly. The floating `]` from step 1 is assumed to belong to the typed `[`. The softchar `]` was inserted too late to be a part of the styling calculation, even though it was correctly styled in the end.

Any clue where the actual softchar insertion takes place in the code?

**Comment by ervumlens** [comment]
Two other observations: 

1. Replacing [the async call to doUpdateUI](https://github.com/Komodo/KomodoEdit/blob/8865f7249de2715c648b6aebc37b0f00b6466214/src/chrome/komodo/content/bindings/views-buffer.p.xml#L1169) with the sync equivalent did not correct this problem, even though it fixed #396 (potentially breaking who-knows-what, but that's a different story). 
2. Lengthening the timeout delay to seconds in the async call *delayed the auto-insertion of the softchar* and did not correct this problem.

**Comment by ervumlens** [comment]
The issue becomes more apparent after changing `bracebad` styling from the default values:

![image](https://cloud.githubusercontent.com/assets/13005338/8626776/14b9f4fa-270c-11e5-9314-65cd37bbfee4.png)

This is probably the style mismatch that cgchoffman noticed, i.e., `bracebad` color on the left,  `operator` color on the right.

There are two underlying problems.
1. Brace highlighting occurs before the softchar is inserted ([line 1109](https://github.com/Komodo/KomodoEdit/blob/master/src/chrome/komodo/content/bindings/views-buffer.p.xml#L1109) vs [line 1146](https://github.com/Komodo/KomodoEdit/blob/master/src/chrome/komodo/content/bindings/views-buffer.p.xml#L1146)).
2. The softchar is not styled until some time after the `doUpdateUI` call ends. Brace highlighting depends on style matching, which requires accurate styling.

Problem 1 is easy to fix. Problem 2 requires firing a restyle on the inserted text.

Any thoughts on how to trigger a restyle?

**Comment by ervumlens** [comment]
Here's another side-effect of this problem. 

I started writing a regex to find paired braces (for reasons unrelated to this bug, thanks for asking :unamused: ). I started out with this: `/\[\]/`. No surprises here. But then I add another `[` immediately after the `[`:

![image](https://cloud.githubusercontent.com/assets/13005338/8631611/a6d50828-2740-11e5-818e-a58b9450e91d.png)


The softchar `]` inserts as expected. The highlighting borks because the softchar is unstyled noise.
